### PR TITLE
Compute cashflows

### LINKED ISSUE #427: Write Tests on Cashflower

### PR DESCRIPTION
**Pull request by sbatka**
- correct Compute methods, 
- including adding "Custom" (corresponds to simple taking of provided values, as in IFRS17 Calc Engine)
- add tests of the "Compute" methods
- update documentation & cleanup

### CODE REVIEW

**Comment by sbatka** [comment]
@dcolleoni I now merged your changes, including last PR (ImproveInit). the tests pass.

File: fpna-template/Report/Solvency.ipynb
Comment (dcolleoni) [inline fpna-template/Report/Solvency.ipynb:1]: Is this on purpose?

File: fpna/Import/Scopes.ipynb [hunk 1]
```diff
@@ -63,6 +68,7 @@
                 "\n",
                 "\npublic interface CompoundShape : ComputeBasicCashflow {",
                 "\n    double[] ComputeBasicCashflow.Values => Identity.InitialValue.GetCompoundValues(Identity.Shift, Identity.Length, Identity.Values);",
+                "\n",
```
Comment (dcolleoni) [inline fpna/Import/Scopes.ipynb]: remove extra line

File: fpna/Tests/ExtensionMethodsTest.ipynb [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                "<p style=\"font-weight:bold;\"> <span style=\"font-size: 36px\"> Extension Methods Test </span> </p>"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "#!import \"../CalculationEngine\""
```
Comment (dcolleoni) [inline fpna/Tests/ExtensionMethodsTest.ipynb]: I think you only need the ExtensionMethods

File: fpna/Utils/ExtentionMethods.ipynb
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb:1]: Pls correct the typo in the fileName. "Extension"

File: fpna/Utils/ExtentionMethods.ipynb [hunk 3]
```diff
@@ -19,7 +19,28 @@
...
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public static double[] GetConstantValues(this double initialValue, int shift, int length){",
+                "\n",
+                "\n    var constant = new double[shift + length];",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: I find confusing having a length parameter which does not controls the length of the array.
Didn't we say that the array should indeed have Lenght = length?

File: fpna/Utils/ExtentionMethods.ipynb [hunk 4]
```diff
@@ -40,14 +64,12 @@
...
-                "\n    for (var i = shift; i < length; i++)",
-                "\n        if (i == shift)",
-                "\n            linear[i] = initialValue;",
-                "\n        else",
-                "\n            linear[i] = linear[i-1] * values[i - shift];",
+                "\n",
+                "\n    var linear = new double[shift + length];",
                 "\n    ",
+                "\n    for (var i = 0; i < length; i++)",
+                "\n        linear[shift + i] =  initialValue + values.Take(i + 1).Aggregate((x, y) => x + y);",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: .Sum() instead of Aggregate

File: fpna/Utils/ExtentionMethods.ipynb [hunk 5]
```diff
@@ -70,14 +95,12 @@
...
-                "\n    ",
-                "\n    for (var i = shift; i < length; i++)",
-                "\n        if (i == shift)",
-                "\n            compound[i] = initialValue;",
-                "\n        else",
-                "\n            compound[i] = compound[i-1] * (1d + values[i - shift]);",
                 "\n    ",
+                "\n    var compound = new double[shift + length];",
+                "\n",
+                "\n    for (var i = 0; i < length; i++)",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: var i = 0 or var i = shift?

File: fpna/Utils/ExtentionMethods.ipynb [hunk 6]
```diff
@@ -59,8 +81,11 @@
             "cell_type": "markdown",
             "source": [
                 "## Compute Compounded Cashflow",
+                "\nThis method provides a geometrically increasing or decreasing cash flow.",
                 "\n",
-                "\nTODO: Add Short description"
+                "\nThe **amount** of the cash flow in the first period is equal to \"**Initial Value**\" * (1 + **Value**(t)). In subsequent periods, the cash flow is equal to the previous cash flow multiplied by (1 + **Value**(t)).",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: why (1+Value(t)) and not just Value(t)?
if you want the cashflow to go to:
- 0, you need to provide -1 (0 would be more intuitive)
- -previousValue, you need to provide -2 (-1 would be more intuitive)

File: fpna/Utils/ExtentionMethods.ipynb [hunk 7]
```diff
@@ -70,14 +95,12 @@
...
-                "\n    for (var i = shift; i < length; i++)",
-                "\n        if (i == shift)",
-                "\n            compound[i] = initialValue;",
-                "\n        else",
-                "\n            compound[i] = compound[i-1] * (1d + values[i - shift]);",
                 "\n    ",
+                "\n    var compound = new double[shift + length];",
+                "\n",
+                "\n    for (var i = 0; i < length; i++)",
+                "\n        compound[shift + i] = initialValue * values.Take(i + 1).Select(x => 1 + x).Aggregate((x, y) => x * y);",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: If the formulas across methods are similar I would immediately extract it. 
You can provide a Func to control the operation (sum or multiplication): 
image

File: fpna/Utils/ExtentionMethods.ipynb [hunk 8]
```diff
@@ -108,9 +133,9 @@
         {
             "cell_type": "code",
             "source": [
-                "public static double[] GetBoundedRandomCashflow(this int shift, int length, double max, double min){",
+                "public static double[] GetBoundedRandomValues(this int shift, int length, double max, double min){",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: I think this method should include the InitialValue. 
I think the BoundedCashflow has that property

File: fpna/Tests/ExtensionMethodsTest.ipynb (hunk 2 above)
Comment (sbatka) [reply fpna/Tests/ExtensionMethodsTest.ipynb]: I am using "Precision" that is defined in ifrs CE so I change this to point there now.

File: fpna/Utils/ExtentionMethods.ipynb [hunk 9]
```diff
@@ -121,6 +146,117 @@
...
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public static double GetCorrelatedValue(double[] r1, double[] r2, double[,] correlation){",
+                "\n",
+                "\n    if (r1.Length != r2.Length)",
+                "\n        throw new System.Exception(\"Vectors of unequal lenghts!\");",
```
Comment (dcolleoni) [inline fpna/Utils/ExtentionMethods.ipynb]: no exception.
check out GetValidElement()

File: fpna/Utils/ExtentionMethods.ipynb (hunk 5 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: changed to be var i = shift

File: fpna/Utils/ExtentionMethods.ipynb (hunk 3 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: changed the meaning of "Length", including the description.

File: fpna/Utils/ExtentionMethods.ipynb (hunk 4 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: I prefer to keep the "Aggregate" to be able to move to usage "EvaluateFunctionElementByElement" method later

File: fpna/Utils/ExtentionMethods.ipynb (hunk 3 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: changed the meaning of length

File: fpna/Utils/ExtentionMethods.ipynb (hunk 6 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: prefer to keep 1+  Value, to allow providing growth rate (e.g. interest rate values, will provide a growth of an amount over time)

File: fpna/Utils/ExtentionMethods.ipynb (hunk 9 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: removed exception

File: fpna/Utils/ExtentionMethods.ipynb (hunk 8 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: - I think this should give purely random number. 
- we will add later some statistical distributions where the initialValue will be mean. but for now I prefer to keep it now.

File: fpna/Utils/ExtentionMethods.ipynb (hunk 7 above)
Comment (sbatka) [reply fpna/Utils/ExtentionMethods.ipynb]: I will add a TODO here

File: fpna-template/Report/Solvency.ipynb
Comment (Gentleman2023) [reply fpna-template/Report/Solvency.ipynb:1]: yes, its duplicated in Earnings

Comment (sbatka) [reply fpna-template/Report/Solvency.ipynb:1]: yes its replicated in Earnings

**Review by dcolleoni — APPROVED** [review]

**Pull request merged**